libtool/run.tcc

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2015-2021 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZETH_LIBTOOL_RUN_TCC__
+#define __ZETH_LIBTOOL_RUN_TCC__
+
+#include "libtool/run.hpp"
+
+namespace libtool
+{
+
+namespace internal
+{
+
+template<typename GlobalOptionsT>
+void print_usage(
+    const char *const *argv,
+    const po::options_description &options,
+    const std::map<std::string, subcommand<GlobalOptionsT> *> &subcommands)
+{
+    std::cout << "Usage:\n"
+              << "  " << argv[0]
+              << " [<options>] <command> <command-arguments> ...\n\n"
```

Comment:
Why aren't the options after the command: `zeth_tool verify <my_options>` (`<my options>` being options of `zeth_tool` or options of `verify` (i.e the command))?

Reply:
This part refers to the global options, which appear before the command name.  The arguments for specific commands (`<command-arguments>`) are described in the usage for that command.

Reply:
I see, thanks @dtebbs 

Here's a detail w.r.t the usage message. We currently have:
```
$ zeth
Usage: zeth [OPTIONS] COMMAND [ARGS]...

Options:
  --eth-network TEXT         Ethereum RPC endpoint, network or config file
                             (default: 'eth-network' if it exists, otherwise
                             'ganache')
  --prover-server TEXT       Prover server endpoint (default=localhost:50051)
  --prover-config-file TEXT  Prover config file (default=prover-config.cache)
  --instance-file TEXT       Instance file (default=zeth-instance)
  --address-file TEXT        Zeth zecret address file (default=zeth-
                             address.priv)
  --wallet-dir TEXT          Wallet directory (default=./wallet)
  --help                     Show this message and exit.

Commands:
  help*                 Print help and exit
  deploy                Deploy the zeth contracts and record the...
  gen-address           Generate a new Zeth secret key and public address
  get-verification-key  Command help text.
  ls-commits            List all commitments in the Zeth mixer contract
  ls-notes              List the set of notes owned by this wallet
  mix                   Generic mix function
  sync                  Attempt to retrieve new notes for the key in...
  wait                  Wait for a mix transaction and dump all log data.
Error: no command specified

$ zeth-tool                         ✘ 1
error: no command specified
Usage:
  ./build/zeth_tool/zeth-tool [<options>] <command> <command-arguments> ...

Global options:
  -h [ --help ]         This help
  -v [ --verbose ]      Verbose output

Commands:
  dump-proof    Print an existinng proof
  prove         Generate proof given proving key and assignment
  verify        Verify proof against verification key and primary input


$ zeth-helper
Usage: zeth-helper [OPTIONS] COMMAND [ARGS]...

Options:
  --eth-network TEXT  Ethereum RPC endpoint, network or config file (default:
                      'eth-network' if it exists, otherwise 'ganache')
  --help              Show this message and exit.

Commands:
  help*                     Print help and exit
  eth-fund                  Fund an address.
  eth-gen-address           Locally generate a new Ethereum private key and...
  eth-gen-network-config    Generate a network config file.
  eth-get-balance           Command to get the balance of specific...
  eth-get-contract-address  Extract the address from a contract instance...
  eth-send                  Send Ether from the local eth-addr to a...
  token-approve             Approve the mixer to spend some amount of...
Error: no command specified
```


As such, I'd change the "usage string" of `zeth-tool` from `zeth-tool [<options>] <command> <command-arguments> ...` to `zeth-tool [OPTIONS] COMMAND [ARGS]...` for the sake of interface consistency across our various CLIs